docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+> out to complicate builds due to varying link-time requirements. As we only
+> intend to use D-Bus for pub-sub messaging ("signals"), we may instead consider
+> to provide a purpose-built in-memory pub-sub service.
+
+
+### Signatures
+
+Each process which needs to produce signatures using the `radicle-link` _device
+key_, SHALL delegate this to an agent conforming to the [SSH agent
+protocol][ssh-agent] (section 4.5.).
+
+Key access restrictions thus depend on user configuration: the agent could be
+configured to prompt for a passphrase after a timeout elapses (or never),
+require presence of a hardware security token, or utilise platform
+authenticators such as TouchID&trade;.
+
+Key generation SHOULD, however, be performed using `radicle-link`-supplied
+tooling building on the [ed25519-zebra] library, which prevents certain weaker
+keys from being generated.
+
+Signature verification MUST be performed using the [ed25519-zebra] library, and
+thus MUST NOT be delegated to the agent.
+
+Note that (hypothetical) applications which require a large number of signatures
+to be generated (such as data migration tools) may find `ssh-agent` throughput
+insufficient, and require the raw key material to be supplied directly. This
+means that long-term secure key storage remains a concern of the `radicle-link`
+application suite.
+
+
+### Process Orchestration
+
+All daemon processes SHALL be started on-demand using the platform socket
+activation protocol. Unless otherwise noted, daemons SHALL terminate after a
+configured period of idle time (in the order of 10s of minutes). The platform
+process manager MUST NOT restart socket-activated daemon processes if they exit
+with a success status (except by another request for the socket).
+
+All daemon processes SHALL run with user privileges (typically the logged-in
+user). Further confinement (e.g. SELinux) is beyond the scope of this document.
+
+
+### IPC
+
+All communication with daemon processes SHALL occur over UNIX domain sockets in
+`SOCK_STREAM` mode. The socket files MUST be stored in a directory only
+accessible by the logged-in user (typically `$XDG_RUNTIME_DIR`), and have
+permissions `0700` by default. Per-service socket paths are considered "well
+known".
+
+RPC calls over those sockets use [CBOR] for their payload encoding. As
+incremental decoders are not available on all platforms, CBOR-encoded messages
+shall be prepended by their length in bytes, encoded as a 32-bit unsigned
+integer in network byte order.
+
+Our security model does not include an attacker which is able to intercept
+traffic on local UNIX sockets. We also do not consider the possibility of
+substituting a service executable with a malicious one listening on the "well
+known" socket for that service. Consequently, we do not require transport-layer
+security nor verification of the service's identity.
+
+We do, however, require that client applications requesting privileged
```

Comment:
Admittedly, this is somewhat paranoid: as long as the service socket and `SSH_AUTH_SOCK` have the same permissions, they are both in the same “security domain”, and this would not be needed. I am, however, sure that reasons will be found to relax the service socket permissions =]